File: NewsAlarm/app/src/main/java/com/alarm/newsalarm/newsmanager/NewsArticleCrawler.java
package com.alarm.newsalarm.newsmanager;

import android.util.Pair;

import com.alarm.newsalarm.utils.LogUtil;

import org.json.JSONException;
import org.json.JSONObject;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.concurrent.Callable;
import java.util.concurrent.CancellationException;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.util.concurrent.RejectedExecutionException;
import java.util.concurrent.TimeUnit;

class NewsArticleCrawler {

    private static final String CLASS_NAME = "NewsArticleCrawler";

    private final ExecutorService service = Executors.newFixedThreadPool(5);
    private final List<Callable<Pair<String, String>>> excutorList = new ArrayList<>();
    private List<Future<Pair<String, String>>> futureList;

    Pair<ArrayList<String>, ArrayList<String>> crawl(Set<JSONObject> jsonArticleSet) {
        for (JSONObject article : jsonArticleSet) {
            excutorList.add(new CrawlingExecutor(article));
        }
        return crawlAsync();
    }

    private Pair<ArrayList<String>, ArrayList<String>> crawlAsync() {
        try {
            futureList = service.invokeAll(excutorList, 2, TimeUnit.SECONDS);
            return getCompletedArticleList();
        } catch (InterruptedException | RejectedExecutionException | NullPointerException e) {
            LogUtil.logE(CLASS_NAME, "crawlAsync", e);
        }
        return new Pair<>(new ArrayList<>(), new ArrayList<>());
    }

    private Pair<ArrayList<String>, ArrayList<String>> getCompletedArticleList() {
        ArrayList<String> titleList = new ArrayList<>();
        ArrayList<String> bodyList = new ArrayList<>();
        for (int i = 0; i < futureList.size(); i++) {
            try {
                Pair<String, String> article = futureList.get(i).get();
                if ("".equals(article.first) || "".equals(article.second)) {
                    continue;
                }
                titleList.add(article.first);
                bodyList.add(article.second);
            } catch (ExecutionException | InterruptedException | CancellationException e) {
                LogUtil.logE(CLASS_NAME, "getCompletedArticleList", e);
            }
        }
        return new Pair<>(titleList, bodyList);
    }

    void release() {
        service.shutdown();
    }

    record CrawlingExecutor(JSONObject jsonArticle) implements Callable<Pair<String, String>> {

        @Override
        public Pair<String, String> call() {
            return new Pair<>(getArticleTitle(), getArticleBody());
        }

        private String getArticleTitle() {
            try {
                return jsonArticle.getString("title").replaceAll("([\\[(<&](.*?)[;>)\\]])", "");
            } catch (JSONException | NullPointerException e) {
                LogUtil.logE(CLASS_NAME, "getArticleTitle", e);
            }

            return "";
        }

        private String getArticleBody() {
            try {
                return modifyContent(crawlPassage(jsonArticle.getString("link")));
            } catch (JSONException | NullPointerException e) {
                LogUtil.logE(CLASS_NAME, "getArticleBody", e);
            }

            return "";
        }

        private String modifyContent(Element e) {
            if (e == null) {
                return "";
            }

            String content = e.text().replaceAll("([\\[(<&](.*?)[;>)\\]])", "");
            final char[] exceptChars = {'#', '※', '▶', 'ⓒ'};
            for (char exceptChar : exceptChars) {
                int index = content.indexOf(exceptChar);
                if (index != -1) {
                    content = content.substring(0, index);
                    break;
                }
            }
            return content;
        }

        private Element crawlPassage(String link) {
            try {
                Document doc = Jsoup.connect(link).get();
                Element e = doc.selectFirst("#dic_area");
                if (e == null) {
                    e = doc.selectFirst("#articeBody");
                }
                if (e == null) {
                    e = doc.selectFirst("#newsEndContents");
                }

                int num = e.childrenSize();
                for (int i = num - 1; i >= 0; i--) {
                    String tag = e.child(i).tagName();
                    if ((!tag.equals("font") && !tag.equals("span") && !tag.equals("b"))
                        || e.child(i).className().equals("end_photo_org")) {
                        e.child(i).remove();
                    }
                }
                LogUtil.logI(CLASS_NAME, "crawlPassage", "crawling completed!");

                return e;
            } catch (IOException | NullPointerException e) {
                LogUtil.logE(CLASS_NAME, "crawlPassage", e);
            }

            return null;
        }
    }
}

File: NewsAlarmKt/app/src/main/java/com/alarm/newsalarmkt/newsmanager/NewsArticleCrawler.kt
package com.alarm.newsalarmkt.newsmanager

import android.util.Pair
import com.alarm.newsalarmkt.utils.LogUtil
import org.json.JSONObject
import org.jsoup.Jsoup
import org.jsoup.nodes.Element
import java.util.concurrent.Callable
import java.util.concurrent.Executors
import java.util.concurrent.Future
import java.util.concurrent.TimeUnit

internal class NewsArticleCrawler {

    private val service = Executors.newFixedThreadPool(5)
    private val executorList = ArrayList<Callable<Pair<String, String>>>()
    private lateinit var futureList: MutableList<Future<Pair<String, String>>>

    fun crawl(jsonArticleSet: MutableSet<JSONObject>): Pair<ArrayList<String>, ArrayList<String>> {
        for (article in jsonArticleSet) {
            executorList.add(CrawlingExecutor(article))
        }

        return crawlAsync()
    }

    private fun crawlAsync(): Pair<ArrayList<String>, ArrayList<String>> {
        try {
            futureList = service.invokeAll(executorList, 2, TimeUnit.SECONDS)
            return completedArticleList
        } catch (e: Exception) {
            LogUtil.logE(CLASS_NAME, "crawlAsync", e)
        }

        return Pair(ArrayList(), ArrayList())
    }

    private val completedArticleList: Pair<ArrayList<String>, ArrayList<String>>
        get() {
            val titleList = ArrayList<String>()
            val bodyList = ArrayList<String>()
            for (i in futureList.indices) {
                try {
                    val article = futureList[i].get()
                    if ("" == article.first || "" == article.second) {
                        continue
                    }
                    titleList.add(article.first)
                    bodyList.add(article.second)
                } catch (e: Exception) {
                    LogUtil.logE(CLASS_NAME, "getCompletedArticleList", e)
                }
            }
            return Pair(titleList, bodyList)
        }

    fun release() {
        service.shutdown()
    }

    private data class CrawlingExecutor(val jsonArticle: JSONObject)
        : Callable<Pair<String, String>> {

        override fun call(): Pair<String, String> {
            return Pair(articleTitle, articleBody)
        }

        private val articleTitle: String
            get() {
                try {
                    return jsonArticle
                        .getString("title")
                        .replace("([\\[(<&](.*?)[;>)\\]])".toRegex(), "")
                } catch (e: Exception) {
                    LogUtil.logE(CLASS_NAME, "getArticleTitle", e)
                }

                return ""
            }

        private val articleBody: String
            get() {
                try {
                    return modifyContent(crawlPassage(jsonArticle.getString("link")))
                } catch (e: Exception) {
                    LogUtil.logE(CLASS_NAME, "getArticleBody", e)
                }

                return ""
            }

        private fun modifyContent(e: Element?): String {
            if (e == null) {
                return ""
            }

            var content = e.text().replace("([\\[(<&](.*?)[;>)\\]])".toRegex(), "")
            val exceptChars = charArrayOf('#', '※', '▶', 'ⓒ')
            for (exceptChar in exceptChars) {
                val index = content.indexOf(exceptChar)
                if (index != -1) {
                    content = content.take(index)
                    break
                }
            }
            return content
        }

        private fun crawlPassage(link: String): Element? {
            try {
                val doc = Jsoup.connect(link).get()
                val e = doc.selectFirst("#dic_area")
                    ?: doc.selectFirst("#articeBody")
                    ?: doc.selectFirst("#newsEndContents")
                    ?: return null

                val num = e.childrenSize()
                for (i in num - 1 downTo 0) {
                    val tag = e.child(i).tagName()
                    if (((tag != "font") && (tag != "span") && (tag != "b"))
                        || e.child(i).className() == "end_photo_org"
                    ) {
                        e.child(i).remove()
                    }
                }
                LogUtil.logI(CLASS_NAME, "crawlPassage", "crawling completed!")

                return e
            } catch (e: Exception) {
                LogUtil.logE(CLASS_NAME, "crawlPassage", e)
            }

            return null
        }
    }

    companion object {
        private const val CLASS_NAME = "NewsArticleCrawler"
    }
}
